plugins/alembic/module/vtkF3DAlembicReader.cxx

```diff
@@ -20,51 +22,334 @@
+        duplicated_data._indices[i].resize(
+          this_face_vertex_count, Alembic::Abc::V3i(-9999, -9999, -9999));
       }
 
-      vtkNew<vtkIdTypeArray> offsets;
-      vtkNew<vtkIdTypeArray> connectivity;
-      offsets->SetNumberOfTuples(counts->size() + 1);
-      connectivity->SetNumberOfTuples(indices->size());
+      // Points
+      {
+        V3fContainer P_v3f;
+        int P_running_index = 0;
+        for (size_t i = 0; i < face_count; i++)
+        {
+          auto this_face_vertex_count = original_data._indices[i].size();
+          for (size_t j = 0; j < this_face_vertex_count; j++)
+          {
+            Alembic::Abc::V3f original_position =
+              original_data._attributes.at("P")[original_data._indices[i][j].x];
+            P_v3f.emplace_back(original_position);
+            duplicated_data._indices[i][j].x = P_running_index;
+            P_running_index++;
+          }
+        }
+
+        duplicated_data._attributes.insert(AttributesContainer::value_type("P", P_v3f));
+      }
 
-      offsets->SetTypedComponent(0, 0, 0);
-      for (size_t i = 0; i < counts->size(); i++)
+      // UV
+      if (have_uv)
       {
-        offsets->SetTypedComponent(i + 1, 0, offsets->GetTypedComponent(i, 0) + counts->get()[i]);
+        V3fContainer uv_v3f;
+        int uv_running_index = 0;
+
+        for (size_t i = 0; i < face_count; i++)
+        {
+          auto this_face_vertex_count = original_data._indices[i].size();
+          for (size_t j = 0; j < this_face_vertex_count; j++)
+          {
+            Alembic::Abc::V3f original_uv =
+              original_data._attributes.at("uv")[original_data._indices[i][j].y];
+            uv_v3f.emplace_back(original_uv);
+            duplicated_data._indices[i][j].y = uv_running_index;
+            uv_running_index++;
+          }
+        }
+
+        duplicated_data._attributes.insert(AttributesContainer::value_type("uv", uv_v3f));
       }
 
-      for (size_t i = 0; i < indices->size(); i++)
+      // Normal
+      if (have_N)
       {
-        connectivity->SetTypedComponent(i, 0, indices->get()[i]);
+        V3fContainer N_v3f;
+        int N_running_index = 0;
+
+        for (size_t i = 0; i < face_count; i++)
+        {
+          auto this_face_vertex_count = original_data._indices[i].size();
+          for (size_t j = 0; j < this_face_vertex_count; j++)
+          {
+            Alembic::Abc::V3f original_N =
+              original_data._attributes.at("N")[original_data._indices[i][j].z];
+            N_v3f.emplace_back(original_N);
+            duplicated_data._indices[i][j].z = N_running_index;
+            N_running_index++;
+          }
+        }
+
+        duplicated_data._attributes.insert(AttributesContainer::value_type("N", N_v3f));
       }
+    }
+    else
+    {
+      duplicated_data = original_data;
+    }
+  }
+
+  void WriteVTP(const IntermediateGeometry& data, vtkPolyData* polydata)
```

Comment:
`VTP` is the file format, not the internal structure, let's rename that function to something like `FillPolyData`